moonsense/cli/download.py

```diff
@@ -0,0 +1,191 @@
+from queue import Empty
+
+from google.protobuf.json_format import MessageToJson
+from retry import retry
+
+from moonsense.client import Client
+
+MISSING_GROUP_ID = "missing-group-id"
+GRACE_PERIOD = 2
+KILL_PERIOD = 30
+
+
+moonsense_client = Client()
+
+@retry(Exception, tries=3, delay=0)
+def download_data_into_folder(datadir, with_group_id, session_id):
+    session = None
+    try:
+        session = moonsense_client.describe_session(session_id)
+    except Exception as e:
+        print("Error encountered while describing session", e)
+        raise e
+    group_id = session.client_session_group_id
+    if group_id is None or len(group_id) == 0:
+        group_id = MISSING_GROUP_ID
+    
+    session_as_json = MessageToJson(session)
+
+    created_at = datetime.fromtimestamp(session.created_at.seconds).date()
+    formatted_created_at = created_at.strftime("%Y-%m-%d")
+
+    session_dir_path = os.path.join(datadir, formatted_created_at, session.session_id)
+    if with_group_id:
+        session_dir_path = os.path.join(datadir,
+            created_at.strftime("%Y-%m-%d"),
+            group_id,
+            session.session_id)
+
+    metadata_path = os.path.join(session_dir_path, "metadata.json")
+    with open(metadata_path, "w") as metadata_file:
+      metadata_file.write(session_as_json)
+
+    raw_sealed_bundles_path = os.path.join(session_dir_path, "raw_sealed_bundles.json")
+    pcap_path = os.path.join(session_dir_path, "packets.pcap")
+
+    try:
+        print(f"Downloading session {session_id} from {formatted_created_at}")
+        moonsense_client.download_session(session_id, raw_sealed_bundles_path)
+        if 'packet' in session.counters:
+            moonsense_client.download_pcap_data(session_id, pcap_path)
+    except Exception as e:
+        print("Error encountered while downloading session", e)
+        raise e
+
+
+def download_session(queue, stop_event, with_group_id):
+    signal.signal(signal.SIGINT, signal.SIG_IGN)
+
+    localdir = os.getcwd()
+    datadir = os.path.join(localdir, "data")
+
+    while True:
+        if stop_event.is_set():
+            break
+
+        try:
+            msg = queue.get(timeout=.01)
+        except Empty:
+            # Run next iteration of loop
+            continue
+
+        if msg is None:
+            queue.task_done()
+
+        try:
+            session_id = msg
+            download_data_into_folder(datadir, with_group_id, session_id)
+        finally:
+            queue.task_done()
+
+
+def fetch_group_id(session):
+    return session.client_session_group_id
+
+
+def handler(signalname):
+    def wrapper_handler(signal_received, frame):
+        raise KeyboardInterrupt(f"{signalname} received")
+    return wrapper_handler
+
+def run_download(until: str, since: str, labels: list[str], with_group_id: bool = False):
+    """
+    Download sessions from a project.
+
+    :param until: Date in the YYYY-MM-DD format until the session data should be included.
```

Comment:
I think passing an optional output path in here would be preferable. Outputting to cwd()/data is fine as a default but with it being a CLI tool, before forced to `cd` to the parent of the output dir to run it is a bit weird.

Reply:
Sure.